File: src/Terrabuild.EventQueue/EventQueue.cs
using System.Runtime.ExceptionServices;
using System.Threading.Channels;

namespace Terrabuild.EventQueue;

public enum Priority {
    Normal,
    Background
}

public interface IEventQueue : IDisposable {
    public void Enqueue(Priority kind, Action action);
    public ExceptionDispatchInfo? WaitCompletion();
}

public sealed class EventQueue : IEventQueue {
    private readonly int _maxConcurrency;
    private readonly int _backgroundMaxConcurrency;

    private readonly Channel<WorkItem> _normal;
    private readonly Channel<WorkItem> _background;

    private readonly Lock _gate = new();

    private volatile bool _started;
    private int _pending;

    private ExceptionDispatchInfo? _lastError;

    private Task[] _workers = [];
    private readonly TaskCompletionSource _drained =
        new(TaskCreationOptions.RunContinuationsAsynchronously);

    private readonly CancellationTokenSource _cts = new();

    private readonly struct WorkItem(Priority kind, Action run) {
        public Priority Kind { get; } = kind;
        public Action Run { get; } = run;
    }

    public EventQueue(int maxConcurrency) {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(maxConcurrency);

        _maxConcurrency = maxConcurrency;
        _backgroundMaxConcurrency = 4 * maxConcurrency;

        var options = new UnboundedChannelOptions {
            SingleReader = false,
            SingleWriter = false,
            AllowSynchronousContinuations = true
        };

        _normal = Channel.CreateUnbounded<WorkItem>(options);
        _background = Channel.CreateUnbounded<WorkItem>(options);
    }

    public void Enqueue(Priority kind, Action action) {
        ArgumentNullException.ThrowIfNull(action);

        // Once we have an error, normal work is dropped.
        if (kind == Priority.Normal && Volatile.Read(ref _lastError) != null) {
            return;
        }

        Interlocked.Increment(ref _pending);

        var work = new WorkItem(kind, action);
        var writer = kind == Priority.Normal ? _normal.Writer : _background.Writer;

        // Very fast for unbounded channels
        if (!writer.TryWrite(work)) {
            // Extremely rare for unbounded, but safe fallback
            _ = writer.WriteAsync(work, _cts.Token).AsTask().ContinueWith(_ => {
                // If WriteAsync failed, mark as complete so WaitCompletion doesn't hang.
                DecrementPending();
            }, TaskScheduler.Default);
        }
    }

    public ExceptionDispatchInfo? WaitCompletion() {
        EnsureStarted();

        // Wait until everything accepted has finished
        _drained.Task.GetAwaiter().GetResult();

        // Now close channels so workers exit
        _normal.Writer.TryComplete();
        _background.Writer.TryComplete();

        // Ensure workers are done
        Task.WaitAll(_workers);

        return _lastError;
    }

    private void EnsureStarted() {
        if (_started) {
            return;
        }

        lock (_gate) {
            if (_started) {
                return;
            }

            _started = true;

            // If nothing is pending at the moment we start, we are already drained.
            if (Volatile.Read(ref _pending) == 0) {
                _drained.TrySetResult();
            }

            _workers = StartWorkers();
        }
    }

    private Task[] StartWorkers() {
        var tasks = new Task[_maxConcurrency + _backgroundMaxConcurrency];
        var i = 0;

        for (; i < _maxConcurrency; i++) {
            tasks[i] = Task.Run(() => WorkerLoop(_normal.Reader), _cts.Token);
        }

        for (; i < tasks.Length; i++) {
            tasks[i] = Task.Run(() => WorkerLoop(_background.Reader), _cts.Token);
        }

        return tasks;
    }

    private async Task WorkerLoop(ChannelReader<WorkItem> reader) {
        try {
            while (await reader.WaitToReadAsync(_cts.Token).ConfigureAwait(false)) {
                while (reader.TryRead(out var item)) {
                    Run(item);
                }
            }
        } catch (OperationCanceledException) {
            // Dispose / cancellation
        } catch (ChannelClosedException) {
            // In case someone ever completes the channel with an error
        }
    }

    private void Run(WorkItem item) {
        try {
            item.Run();
        } catch (Exception ex) {
            // any error sets lastError + clears normal queue (stops normal scheduling)
            TrySetError(ex);
        } finally {
            DecrementPending();
        }
    }

    private void TrySetError(Exception ex) {
        // Fast-path: already failed
        if (Volatile.Read(ref _lastError) != null) {
            return;
        }

        lock (_gate) {
            if (_lastError != null) {
                return;
            }

            _lastError = ExceptionDispatchInfo.Capture(ex);

            // IMPORTANT: do not fault the channel (no ex argument),
            // otherwise WaitToReadAsync will throw.
            _normal.Writer.TryComplete();

            // Drop any already enqueued normal tasks (and fix pending count)
            DrainAndDrop(_normal.Reader);
        }
    }

    private void DrainAndDrop(ChannelReader<WorkItem> reader) {
        while (reader.TryRead(out _)) {
            DecrementPending();
        }
    }

    private void DecrementPending() {
        var value = Interlocked.Decrement(ref _pending);
        if (_started && value == 0) {
            _drained.TrySetResult();
        }
    }

    public void Dispose() {
        _cts.Cancel();

        _normal.Writer.TryComplete();
        _background.Writer.TryComplete();

        try {
            if (_workers.Length > 0) {
                Task.WaitAll(_workers);
            }
        } catch {
            // ignore on dispose
        }

        _cts.Dispose();
    }
}

File: src/Terrabuild.PubSub/EventQueue.fs
namespace Terrabuild.PubSub

open System
open System.Runtime.ExceptionServices
open System.Threading
open System.Threading.Channels
open System.Threading.Tasks
open Lock

type Priority =
    | Normal = 0
    | Background = 1

type IEventQueue =
    inherit IDisposable
    abstract Enqueue: kind: Priority -> action: (unit -> unit) -> unit
    abstract WaitCompletion: unit -> ExceptionDispatchInfo option

[<Struct>]
type private WorkItem =
    val Kind: Priority
    val Run: unit -> unit

    new (kind: Priority, run: unit -> unit) =
        { Kind = kind; Run = run }

type EventQueue(maxConcurrency: int) =
    do
        if maxConcurrency <= 0 then
            invalidArg (nameof maxConcurrency) "maxConcurrency must be > 0"

    let maxConcurrency = maxConcurrency
    let backgroundMaxConcurrency = 4 * maxConcurrency

    let options =
        UnboundedChannelOptions(
            SingleReader = false,
            SingleWriter = false,
            AllowSynchronousContinuations = true
        )

    let normal = Channel.CreateUnbounded<WorkItem>(options)
    let background = Channel.CreateUnbounded<WorkItem>(options)

    // gate is only for start + first error transition, not hot path
    let gate = Lock()

    let mutable started = false
    let mutable pending = 0
    let mutable lastError : ExceptionDispatchInfo option = None

    let workers : Task array ref = ref [||]

    let drained =
        TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously)

    let cts = new CancellationTokenSource()

    let decrementPending () =
        let value = Interlocked.Decrement(&pending)
        if started && value = 0 then
            drained.TrySetResult() |> ignore

    let drainAndDrop (reader: ChannelReader<WorkItem>) =
        let mutable item = Unchecked.defaultof<WorkItem>
        while reader.TryRead(&item) do
            decrementPending()

    let trySetError (ex: exn) =
        // fast path: already failed
        if (Volatile.Read(&lastError) |> Option.isSome) then
            ()
        else
            lock gate (fun () ->
                match lastError with
                | Some _ -> ()
                | None ->
                    lastError <- Some (ExceptionDispatchInfo.Capture ex)

                    // IMPORTANT: do not fault the channel, otherwise WaitToReadAsync throws
                    normal.Writer.TryComplete() |> ignore

                    // Drop already-enqueued normal tasks and fix pending
                    drainAndDrop normal.Reader
            )

    let run (item: WorkItem) =
        try
            try
                item.Run()
            with ex ->
                // any error sets lastError + clears normal queue (stops normal scheduling)
                trySetError ex
        finally
            decrementPending()

    let workerLoop (reader: ChannelReader<WorkItem>) =
        task {
            try
                let token = cts.Token
                let mutable item = Unchecked.defaultof<WorkItem>

                while! reader.WaitToReadAsync(token) do
                    while reader.TryRead(&item) do
                        run item
            with
            | :? OperationCanceledException ->
                ()
            | :? ChannelClosedException ->
                ()
        } :> Task

    let startWorkers () =
        let tasks = Array.zeroCreate<Task> (maxConcurrency + backgroundMaxConcurrency)

        let mutable i = 0
        while i < maxConcurrency do
            tasks[i] <- Task.Run(fun () -> workerLoop normal.Reader, cts.Token)
            i <- i + 1

        while i < tasks.Length do
            tasks[i] <- Task.Run(fun () -> workerLoop background.Reader, cts.Token)
            i <- i + 1

        tasks

    let ensureStarted () =
        if started then
            ()
        else
            lock gate (fun () ->
                if started then
                    ()
                else
                    started <- true

                    // If nothing pending at start, we are already drained
                    if Volatile.Read(&pending) = 0 then
                        drained.TrySetResult() |> ignore

                    workers.Value <- startWorkers ()
            )

    interface IEventQueue with
        member _.Enqueue kind action =
            if isNull (box action) then
                nullArg (nameof action)

            // Once we have an error, normal work is dropped.
            if kind = Priority.Normal && (Volatile.Read(&lastError) |> Option.isSome) then
                ()
            else
                Interlocked.Increment(&pending) |> ignore

                let work = WorkItem(kind, action)

                let writer =
                    if kind = Priority.Normal then normal.Writer else background.Writer

                // Very fast for unbounded channels
                if not (writer.TryWrite work) then
                    // Extremely rare for unbounded, but safe fallback
                    writer.WriteAsync(work, cts.Token).AsTask()
                        .ContinueWith(
                            (fun (_: Task) ->
                                // If WriteAsync failed, mark as complete so WaitCompletion doesn't hang.
                                decrementPending()
                            ),
                            TaskScheduler.Default
                        )
                    |> ignore

        member _.WaitCompletion() =
            ensureStarted ()

            // Wait until everything accepted has finished
            drained.Task.GetAwaiter().GetResult()

            // Close channels so workers exit
            normal.Writer.TryComplete() |> ignore
            background.Writer.TryComplete() |> ignore

            // Ensure workers are done
            Task.WaitAll(workers.Value)

            lastError

        member _.Dispose() =
            cts.Cancel()

            normal.Writer.TryComplete() |> ignore
            background.Writer.TryComplete() |> ignore

            try
                if workers.Value.Length > 0 then
                    Task.WaitAll(workers.Value)
            with
            | _ -> ()

            cts.Dispose()
